pkg/controller/nodes/executor.go

```diff
@@ -353,17 +365,19 @@ func (c *nodeExecutor) handleQueuedOrRunningNode(ctx context.Context, nCtx *node
 		} else {
 			c.metrics.UnknownErrorDuration.Observe(ctx, startTime, endTime)
 		}
+		// When a node fails, we fail the workflow. Independent of number of nodes succeeding/failing, whenever a first node fails,
+		// the entire workflow is failed.
+		if np == v1alpha1.NodePhaseFailing {
```

Comment:
In case of retries, do we move from Running -> Failing -> Running -> _Potentially_ Succeeded? or it never hits Failing?

Reply:
@EngHabu Why would we go from Failing to Running.

If we reach failing, we will move to finalize() and then to Failed?

Reply:
```We can have a case when we never execute any node```
Shouldn't this happen when trying to executing a node, and we should return system error if that happens?

Reply:
@EngHabu In case of retries, the phase will be `NodePhaseRetryableFailure`

Reply:
We can fail before getting to try to run a node... I'm leaning towards what you proposed earlier.. returning a recognizable error, and checking for that at the top level.. I understand it's more work but I think this is an important metric to get right.